creme/creme_core/menu.py

```diff
@@ -129,54 +107,43 @@ def check_permissions(self, user):
     id = 'creme_core-role_switch'
     label = _('Available roles')
     url_name = 'creme_core__switch_role'
 
-    def render(self, context):
+    def get_context(self, context):
+        context = super().get_context(context)
         user = context['user']
 
+        actions = []
+
         if not user.is_superuser:
-            roles = [*user.roles.all()]
             # TODO: user.normalize_roles(roles)?
-            if len(roles) >= 2:
-                selected_id = user.role_id
-                url_name = self.url_name
-                home_url = reverse('creme_core__home')
-
-                return format_html(
-                    '<span class="ui-creme-navigation-title">{label}</span>{roles}',
-                    label=self.label,
-                    roles=format_html_join(
-                        '\n',
-                        '''<button class="{class}" {attr}'''
-                        '''onclick="creme.utils.ajaxQuery('{url}', {{action: 'post'}}).onDone(function(){{ creme.utils.goTo('{home_url}'); }}).start()">'''  # NOQA
-                        '''<div class="marker-{marker_suffix}"></div>{role}'''
-                        '''</button>''',
-                        (
-                            {
-                                'class': 'disabled-role' if role.deactivated_on else '',
-                                'attr': (
-                                    'disabled '
-                                    if role.id == selected_id or role.deactivated_on else
-                                    ''
-                                ),
-                                'url': reverse(url_name, args=(user.id, role.id)),
-                                'home_url': home_url,
-                                'marker_suffix': (
-                                    'selected' if role.id == selected_id else 'not-selected'
-                                ),
-                                'role': role,
-                            } for role in roles
-                        ),
-                    ),
-                )
-
-        return mark_safe(
-            '<span class="ui-creme-navigation-title ui-creme-navigation-empty" />'
-        )
+            roles = user.roles.all()
+
+            if len(roles) > 1:
+                switch_url_name = self.url_name
+                props = {
+                    "options": {
+                        "redirectOnSuccess": reverse("creme_core__home")
+                    }
+                }
+
+                actions = [
+                    {
+                        "label": str(role),
+                        "disabled": role.deactivated_on,
+                        "selected": role.id == user.role_id,
+                        "url": reverse(switch_url_name, args=(user.id, role.id)),
+                        "props": props
+                    } for role in roles
+                ]
+
+        context["roles"] = actions
+        return context
```

Comment:
Would be more elegant to just pass "user" to the context (i presume it's already here) & make all this logic in the template